### DISCUSSION TITLE
Are there any relevant existing networks or examples that make use of edge characteristics?

### DISCUSSION THREAD

**Discussion by boyin8** [pyg-team/pytorch_geometric#6101] [Q&A] [answered]
Hi, I want to use both edge attributes and node attributes to do some link prediction tasks. Does PyG have a relevant example?

Many thanks!

**Comment by rusty1s** [comment, marked as answer]
I don't think we have an example for that, but it should be pretty straightforward to build from an existing link prediction example and just enhance the GNN with edge feature support. Let me know if you need any more guidance :)

**Reply by boyin8** [reply]
Thanks a lot. I check the PyG's GNN cheatsheet and find some GNN architecture that support edge_attr, such as GATconv. Tacking  [link_pred](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py) as an example, if I want to add edge features to predict link, should I replace GCNconv with GATconv? I would greatly appreciate it you could provide some more detailed steps?

By the way, I want to know how PyG deals with isolatde nodes. Does it just keep itself updated with its own embeddings? If so, is it possible that removing such isolated nodes from the graph has no effect on the final training?
Thanks your help again.

**Reply by rusty1s** [reply]
Yes, that would look something like this:
```python
conv = GATConv(in_channels, out_channels, heads=4, edge_dim=num_edge_features)

x = conv(x, edge_index, edge_attr)
```
while everything else stays the same.

For isolated nodes: Yes, in case of isolated nodes, a node will just transform itself based on its own features. Maintaining/removing these isolated nodes has no effect on other nodes in the network.

**Reply by boyin8** [reply]
Got it! It helps me a lot.
Thanks very much.

**Reply by boyin8** [reply]
Hi, 
I would like to ask some more questions about the details of heterogeneous networks with this edge_attr. I notice that in the example, we  have: del data['A', 'rev_B', 'C'].edge_label to delete the reverse link label.
Do I have to do something like the following：
del data['A', 'rev_B', 'C'].edge_attr

Another question, how does the GAT network update its features about the edges? For example, what is the dimension of the edge after the final update? In the source code of GAT:
`if edge_dim is not None:
            self.lin_edge = Linear(edge_dim, heads * out_channels, bias=False,
                                   weight_initializer='glorot')`
Does this mean that if I have an edge feature dimension of 1, heads=4 and out_channels=4, then the final edge feature dimension is 8?
(also if node feature dimension of 1, heads=4 and out_channels=4, final node feature dimension is 4). Is that right?
Thanks very much.

**Reply by rusty1s** [reply]
I don't think you need to delete the `edge_attr` for the reverse edge since you probably also want to take the edge features into account for it.

Note that `GATConv` does not return edge features, it just uses them to learn attention weights. In your case, which `edge_dim=1` and `out_channels=4` and `heads=4`, it will learn 4 * 4 edge features (4 output features for every head).

**Reply by boyin8** [reply]
Very clear, I got it.

**Reply by boyin8** [reply]
Hi, I am back again.

I adopted your suggestion as you mentioned above, and I got some good results but not satisfied. Dut to the consideration of Heterogeneous and Bipartite graph (two different nodes), I find this [bipartite_sage](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/bipartite_sage.py) more suitable for me. 

My question is if it is possible to simply replace the SAGEConv in this example with another network (GATConv, GeneralConv) that support the edge_attr as we discussed above. As I found it added the metapath in this example, I don't know if the message will pass properly if the edge_attr (or edge_weight) is added.

Another question, as the dimension of edge feature is 1, which do you think is more appropriate, considering edge features or edge weights as input to the network?

Many thanks.

**Reply by rusty1s** [reply]
Yes, you should be able to replace the GNN layer, but I am pretty confident that `AddMetaPaths` can only preserve edge weights, not multi-dimensional edge features.

Whether to treat your edge feature as an edge weight or not depends on its semantic meaning. If it defines the strengths of connections, it is better to treat this as an edge weight, otherwise I would stay with treating this as an edge feature.

**Reply by boyin8** [reply]
Thanks for your answer. For the heterogeneous and bipartite graph, which do you think is more effective, using `to_hetero` function to transform the homegeneous into hteteogeneous or `AddMetaPaths` function to train?

**Reply by rusty1s** [reply]
The `to_hetero` function does not convert the graph into a homogeneous one - it is simply a wrapper to convert a homogeneous GNN into its heterogeneous counterpart where GNN layers are duplicated per edge type. As such, both serve different purposes - what works best probably depends on the task.

**Reply by boyin8** [reply]
Thanks:)

**Reply by boyin8** [reply]
Hi, if I want to add my own defined edge weights to the `AddMetaPaths`, how to do that?
I noticed that if I set the `weighted` parameter to `True` (The weight of each metapath edge is computed as the number of metapaths from the start to the end of the metapath edge), which is not what I want.
Thanks.

**Reply by boyin8** [reply]
> Hi, if I want to add my own defined edge weights to the `AddMetaPaths`, how to do that? I noticed that if I set the `weighted` parameter to `True` (The weight of each metapath edge is computed as the number of metapaths from the start to the end of the metapath edge), which is not what I want. Thanks.

Sorry, In `edge_weight = data[edge_type].get('edge_weight', None)`,  does this mean that the weights I defined are already being used?

**Reply by rusty1s** [reply]
There is a `weighted` option in `AddMetaPaths` which you need to turn on.

**Reply by boyin8** [reply]
Thank you for your advice. However, I've come across a problem regarding whether different network structures can be shared. Specifically, the data I generate randomly each time has the same number of nodes as well as the same type. The only thing that differs from one dataset to another is their network structure, i.e. the `edge_index`. For example, I currently have four sets of training data, A,B,C,D. Can I take the trained model from A and test it on data B? (In practice it seems to be possible and the model returns predictions in PyG).

The reason I need this operation is also because each time my nodes are generated randomly. Or do you have any other recommended methods in PyG?

**Reply by rusty1s** [reply]
Yes, a GNN can operate on unseen network structures. For example, you can train on A and B, validate on C and test on D.

**Reply by rusty1s** [reply]
Take a look at https://colab.research.google.com/drive/1xpzn1Nvai1ygd_P5Yambc_oe4VBPK_ZT?usp=sharing
